analytics.go

```diff
@@ -165,53 +171,120 @@ func (r *RedisAnalyticsHandler) Init() {
 
 	analytics.Store.Connect()
 
-	ps := config.Global().AnalyticsConfig.PoolSize
+	ps := r.globalConf.AnalyticsConfig.PoolSize
 	if ps == 0 {
 		ps = 50
 	}
+	log.WithField("ps", ps).Debug("Analytics pool workers number")
 
-	r.AnalyticsPool, err = tunny.CreatePoolGeneric(ps).Open()
-	if err != nil {
-		log.Error("Failed to init analytics pool")
+	r.recordsBufferSize = r.globalConf.AnalyticsConfig.RecordsBufferSize
+	if r.recordsBufferSize < minRecordsBufferSize {
+		r.recordsBufferSize = minRecordsBufferSize // force it to this value
 	}
-}
-
-// RecordHit will store an AnalyticsRecord in Redis
-func (r *RedisAnalyticsHandler) RecordHit(record AnalyticsRecord) error {
-	r.AnalyticsPool.SendWork(func() {
-		// If we are obfuscating API Keys, store the hashed representation (config check handled in hashing function)
-		record.APIKey = storage.HashKey(record.APIKey)
-
-		if config.Global().SlaveOptions.UseRPC {
-			// Extend tag list to include this data so wecan segment by node if necessary
-			record.Tags = append(record.Tags, "tyk-hybrid-rpc")
-		}
+	log.WithField("recordsBufferSize", r.recordsBufferSize).Debug("Analytics records buffer size")
 
-		if config.Global().DBAppConfOptions.NodeIsSegmented {
-			// Extend tag list to include this data so wecan segment by node if necessary
-			record.Tags = append(record.Tags, config.Global().DBAppConfOptions.Tags...)
-		}
+	chanSize := uint64(ps) * r.recordsBufferSize
```

Comment:
We need to be careful here, in cases when detailed recording is turned on payload size can up to few mb for each record. Just want to ensure that such big buffer won't cause memory issues in edge cases.

Reply:
we are passing pointers to channel, once record sent GC can pick and free it. also, I made that default buffer size too big with assumption that pool workers will process faster than sender sends to channel. if we receive much more and block on writing to channel I think think effect will be close to situation when we we try to schedule lots of go-routines. I was thinking about adaptable worker pool size, i.e. if sender can't send to channel for too long (which means channel is not read too fast) it could ask analytics to increase pool of workers. this sounds interesting but we have reidis on other end and redis connection pool with fixed size - it can get slow just because of redis (thats why I started thinking about sending records via pipelines). Also, according to redis - starting from 2.4 we can provide to RPUSH more then one value, we can benefit from that as well.